src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlDecoder.cs

```diff
@@ -0,0 +1,859 @@
+            [CompExactlyDependsOn(typeof(Avx2))]
+            public static unsafe bool TryLoadAvxVector256(ushort* src, ushort* srcStart, int sourceLength, out Vector256<sbyte> str)
+            {
+                AssertRead<Vector256<sbyte>>(src, srcStart, sourceLength);
+                Vector256<ushort> utf16VectorLower = Avx.LoadVector256(src);
+                Vector256<ushort> utf16VectorUpper = Avx.LoadVector256(src + 16);
+
+                if (VectorContainsNonAsciiChar(utf16VectorLower) || VectorContainsNonAsciiChar(utf16VectorUpper))
+                {
+                    str = default;
+                    return false;
+                }
+
+                str = Vector256.Narrow(utf16VectorLower, utf16VectorUpper).AsSByte();
+                return true;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static unsafe bool TryLoadVector128(ushort* src, ushort* srcStart, int sourceLength, out Vector128<byte> str)
+            {
+                AssertRead<Vector128<sbyte>>(src, srcStart, sourceLength);
+                Vector128<ushort> utf16VectorLower = Vector128.LoadUnsafe(ref *src);
+                Vector128<ushort> utf16VectorUpper = Vector128.LoadUnsafe(ref *src, 8);
+                if (VectorContainsNonAsciiChar(utf16VectorLower) || VectorContainsNonAsciiChar(utf16VectorUpper))
+                {
+                    str = default;
+                    return false;
+                }
+
+                str = Vector128.Narrow(utf16VectorLower, utf16VectorUpper);
+                return true;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            [CompExactlyDependsOn(typeof(AdvSimd.Arm64))]
+            public static unsafe bool TryLoadArmVector128x4(ushort* src, ushort* srcStart, int sourceLength,
+                out Vector128<byte> str1, out Vector128<byte> str2, out Vector128<byte> str3, out Vector128<byte> str4)
+            {
+                AssertRead<Vector128<sbyte>> (src, srcStart, sourceLength);
+                var (s11, s12, s21, s22) = AdvSimd.Arm64.LoadVector128x4AndUnzip(src);
+                var (s31, s32, s41, s42) = AdvSimd.Arm64.LoadVector128x4AndUnzip(src + 32);
+
+                if (VectorContainsNonAsciiChar(s11) || VectorContainsNonAsciiChar(s12) ||
+                    VectorContainsNonAsciiChar(s21) || VectorContainsNonAsciiChar(s22) ||
+                    VectorContainsNonAsciiChar(s31) || VectorContainsNonAsciiChar(s32) ||
+                    VectorContainsNonAsciiChar(s41) || VectorContainsNonAsciiChar(s42))
+                {
+                    str1 = str2 = str3 = str4 = default;
+                    return false;
+                }
+
+                str1 = Vector128.Narrow(s11, s31);
+                str2 = Vector128.Narrow(s12, s32);
+                str3 = Vector128.Narrow(s21, s41);
+                str4 = Vector128.Narrow(s22, s42);
```

Comment:
```suggestion
                str1 = Ascii.ExtractAsciiVector(s11, s31);
                str2 = Ascii.ExtractAsciiVector(s12, s32);
                str3 = Ascii.ExtractAsciiVector(s21, s41);
                str4 = Ascii.ExtractAsciiVector(s22, s42);
```

Since the values are already validated to be ASCII, you can use the slightly more efficient helper from
https://github.com/dotnet/runtime/blob/0651ebc74e671cd000a0a0f029100551d4fd9777/src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Utility.cs#L1655
(same for the other helpers on this type)

Reply:
Great, thank you!